### PR TITLE
Expose extension encoding/decoding in Encoder/Decoder

### PR DESCRIPTION
**Pull request by timoha**
Provides low level functionality in case clients want to bypass RegisterExt() and implement their own generic Encode/Decode function (just as EncodeMapLen and EncodeArrayLen do).

### CODE REVIEW

File: decode.go [hunk 1]
```diff
@@ -335,7 +335,8 @@ func (d *Decoder) DecodeInterface() (interface{}, error) {
 		return d.DecodeMap()
 	case codes.FixExt1, codes.FixExt2, codes.FixExt4, codes.FixExt8, codes.FixExt16,
 		codes.Ext8, codes.Ext16, codes.Ext32:
-		return d.extInterface(c)
+		d.s.UnreadByte()
+		return d.extInterface()
```
Comment (vmihailenco) [inline decode.go]: Can you leave this unchanged? We try to not unread anything for better performance.

File: decode.go [hunk 2]
```diff
@@ -387,7 +388,8 @@ func (d *Decoder) DecodeInterfaceLoose() (interface{}, error) {
...
```
Comment (vmihailenco) [inline decode.go]: same here

File: ext.go [hunk 3]
```diff
@@ -50,6 +50,16 @@ func registerExt(id int8, typ reflect.Type, enc encoderFunc, dec decoderFunc) {
 	}
 }
 
+func (e *Encoder) EncodeExtensionHeader(typeId byte, length int) error {
```
Comment (vmihailenco) [inline ext.go]: There is already `RegisterExt` so this should be called `EncodeExtHeader(typeId int8, length int) error` for consistency (note byte -> int8 change).

File: ext.go [hunk 4]
```diff
@@ -123,13 +130,26 @@ func (d *Decoder) parseExtLen(c codes.Code) (int, error) {
 	}
 }
 
-func (d *Decoder) extInterface(c codes.Code) (interface{}, error) {
+func (d *Decoder) DecodeExtentionHeader() (byte, int, error) {
```
Comment (vmihailenco) [inline ext.go]: `DecodeExtHeader() (typeId int8, length int, err error)` please

**Comment by timoha** [comment]
All done, added a sanity test as well.

**Pull request merged**